### ISSUE TITLE
Unable to reproduce the accuracy using the tutorial on graph attention network (GAT)

### ISSUE THREAD

**Issue by amblee0306** [dmlc/dgl#2337]
Hi there,
I m try to implement GAT using the documentation 
https://docs.dgl.ai/en/0.4.x/tutorials/models/1_gnn/9_gat.html

but i don't seem to be able to get the stated accuracy of '83.69±0.529' on the Cora dataset. I do not want to use the GATConv module as I will be adding things on top of it later and it will thus be more transparent if I can implement GAT from the message passing perspective.

I have added in the feature dropout of 0.6, negative slope of 0.2, weight decay of 5e-4, and changed the loss to cross entropy loss. The only thing that i have not added is the attention dropout of 0.6 as I am not sure if I can add it in the reduce_func.

The test accuracy (after 10 runs) that I have obtained is '77.89 +/- 1.28'.
Please do advice me on what I can do to obtain the mentioned accuracy.
The code that I am using is as below:

```
import torch
import torch.nn as nn
import torch.nn.functional as F

class GATLayer(nn.Module):
    def __init__(self, g, in_dim, out_dim):
        super(GATLayer, self).__init__()
        self.g = g
        # equation (1)
        self.fc = nn.Linear(in_dim, out_dim, bias=False)
        # equation (2)
        self.attn_fc = nn.Linear(2 * out_dim, 1, bias=False)
        self.reset_parameters()

    def reset_parameters(self):
        """Reinitialize learnable parameters."""
        gain = nn.init.calculate_gain('relu')
        nn.init.xavier_normal_(self.fc.weight, gain=gain)
        nn.init.xavier_normal_(self.attn_fc.weight, gain=gain)

    def edge_attention(self, edges):
        # edge UDF for equation (2)
        z2 = torch.cat([edges.src['z'], edges.dst['z']], dim=1)
        a = self.attn_fc(z2)
        return {'e': F.leaky_relu(a, negative_slope=0.2)}

    def message_func(self, edges):
        # message UDF for equation (3) & (4)
        return {'z': edges.src['z'], 'e': edges.data['e']}

    def reduce_func(self, nodes):
        # reduce UDF for equation (3) & (4)
        # equation (3)
        alpha = F.softmax(nodes.mailbox['e'], dim=1)
        # equation (4)
        h = torch.sum(alpha * nodes.mailbox['z'], dim=1)
        return {'h': h}

    def forward(self, h):
        # equation (1)
        z = self.fc(h)
        self.g.ndata['z'] = z
        # equation (2)
        self.g.apply_edges(self.edge_attention)
        # equation (3) & (4)
        self.g.update_all(self.message_func, self.reduce_func)
        return self.g.ndata.pop('h')

class MultiHeadGATLayer(nn.Module):
    def __init__(self, g, in_dim, out_dim, num_heads, merge='cat'):
        super(MultiHeadGATLayer, self).__init__()
        self.heads = nn.ModuleList()
        for i in range(num_heads):
            self.heads.append(GATLayer(g, in_dim, out_dim))
        self.merge = merge

    def forward(self, h):
        head_outs = [attn_head(h) for attn_head in self.heads]
        if self.merge == 'cat':
            # concat on the output feature dimension (dim=1)
            return torch.cat(head_outs, dim=1)
        else:
            # merge using average
            return torch.mean(torch.stack(head_outs))

class GAT(nn.Module):
    def __init__(self, g, in_dim, hidden_dim, out_dim, num_heads):
        super(GAT, self).__init__()
        self.layer1 = MultiHeadGATLayer(g, in_dim, hidden_dim, num_heads)
        self.layer2 = MultiHeadGATLayer(g, hidden_dim * num_heads, out_dim, 1)
        self.feat_drop = nn.Dropout(0.6)

    def forward(self, h):
        h = self.feat_drop(h)
        h = self.layer1(h)
        h = F.elu(h)
        h = self.layer2(h)
        return h

import dgl
from dgl.data import CoraGraphDataset

def load_cora_new():
    data = CoraGraphDataset()
    g = data[0]
    features = g.ndata['feat']
    labels = g.ndata['label']
    train_mask = g.ndata['train_mask']
    val_mask = g.ndata['val_mask']
    test_mask = g.ndata['test_mask']
    # add self loop
    g = dgl.remove_self_loop(g)
    g = dgl.add_self_loop(g)
    return g, features, labels, train_mask, test_mask, val_mask

def accuracy(logits, labels):
    _, indices = torch.max(logits, dim=1)
    correct = torch.sum(indices == labels)
    return correct.item() * 1.0 / len(labels)

def evaluate(model, features, labels, mask):
    model.eval()
    with torch.no_grad():
        logits = model(features)
        logits = logits[mask]
        labels = labels[mask]
        return accuracy(logits, labels)

import time
import numpy as np

def train_model():
    g, features, labels, train_mask, test_mask, val_mask = load_cora_new()
    # create the model, 2 heads, each head has hidden size 8
    net = GAT(g,
              in_dim=features.size()[1],
              hidden_dim=8,
              out_dim=7,
              num_heads=8)

    # create optimizer
    optimizer = torch.optim.Adam(net.parameters(), lr=1e-3, weight_decay=5e-4)

    # main loop
    dur = []
    loss_fcn = torch.nn.CrossEntropyLoss()
    for epoch in range(300):
        if epoch >= 3:
            t0 = time.time()

        logits = net(features)
        loss = loss_fcn(logits[train_mask], labels[train_mask])

        optimizer.zero_grad()
        loss.backward()
        optimizer.step()

        if epoch >= 3:
            dur.append(time.time() - t0)

        val_acc = evaluate(net, features, labels, val_mask)

        print("Epoch {:05d} | Loss {:.4f} | Time(s) {:.4f}".format(
            epoch, loss.item(), np.mean(dur)), '| Val_acc: ', val_acc)

    acc = evaluate(net, features, labels, test_mask)
    print("Test Accuracy {:.4f}".format(acc))

train_model()
```

**Comment by yzh119** [comment]
Actually the GAT module is transparent: https://github.com/dmlc/dgl/blob/master/python/dgl/nn/pytorch/conv/gatconv.py and you can modify it to fit your requirement.

The GAT introduced in the tutorial you mentioned is neither standard nor efficient, it's just designed to let user know how to represent Graph Attention as message passing. We have wrapped many component as builtin-functions (such as edge_softmax,  fn.u_add_v, etc) that were highly efficient.

Please let me know if you don't understand the source code of GATConv module.

**Comment by VoVAllen** [comment]
The hidden_dim seems too small in the code you posted

**Comment by amblee0306** [comment]
> The hidden_dim seems bit small in the code you posted. And you may want to tune the weight_decay also

@VoVAllen based on the GAT paper the hidden dim is 8 with 8 heads which results in a total of 64 dim right? The weight decay is also the one mentioned in the paper, so i dont think that should be an issue? Or am i missing something here?

**Comment by amblee0306** [comment]
> Actually the GAT module is transparent: https://github.com/dmlc/dgl/blob/master/python/dgl/nn/pytorch/conv/gatconv.py and you can modify it to fit your requirement.
> 
> The GAT introduced in the tutorial you mentioned is neither standard nor efficient, it's just designed to let user know how to represent Graph Attention as message passing. We have wrapped many component as builtin-functions (such as edge_softmax, fn.u_add_v, etc) that were highly efficient.
> 
> Please let me know if you don't understand the source code of GATConv module.

@yzh119 inside graphconv there are theses lines

```
            el = (feat_src * self.attn_l).sum(dim=-1).unsqueeze(-1)
            er = (feat_dst * self.attn_r).sum(dim=-1).unsqueeze(-1)
            graph.srcdata.update({'ft': feat_src, 'el': el})
            graph.dstdata.update({'er': er})
```

I checked with these lines above the lines on the top and it is shown that feat_src is always the same as feat_dst. Is there a reason for this?

```
            import torch
            if torch.all(torch.eq(feat_dst,feat_src)):
                print('the two tensors are the same')
            else:
                print('different')
```

output
``` 
Epoch 00000 | Time(s) nan | Loss 1.9447 | TrainAcc 0.1857 | ValAcc 0.2840 | ETputs(KTEPS) nan
the two tensors are the same
the two tensors are the same
the two tensors are the same
the two tensors are the same
Epoch 00001 | Time(s) nan | Loss 1.9371 | TrainAcc 0.2071 | ValAcc 0.4880 | ETputs(KTEPS) nan
the two tensors are the same
the two tensors are the same
the two tensors are the same
the two tensors are the same
Epoch 00002 | Time(s) nan | Loss 1.9278 | TrainAcc 0.2857 | ValAcc 0.6280 | ETputs(KTEPS) nan
the two tensors are the same
the two tensors are the same
the two tensors are the same
the two tensors are the same
Epoch 00003 | Time(s) 0.0617 | Loss 1.9150 | TrainAcc 0.4571 | ValAcc 0.6860 | ETputs(KTEPS) 215.07
the two tensors are the same
the two tensors are the same
the two tensors are the same
the two tensors are the same
```

I still prefer the message parsing method if possible cause I want to get the next hop neighbors making it a total of two hops but i am not sure how to do so using gatconv module. It is in this sense that I meant that it is not transparent. Please do enlighten me.

**Comment by amblee0306** [comment]
@yzh119 is it possible to run this tutorial's (https://docs.dgl.ai/en/0.4.x/tutorials/models/1_gnn/9_gat.html) GAT model and get an accuracy of 83% regardless of the efficiency. It doesn't make sense to me that it is not possible to do so.

**Comment by yzh119** [comment]
@amblee0306 the lines you mentioned are used for GAT on bipartite graphs.

To be more concise, a simplified logic of GAT looks like the following:
```python
    def forward(self, graph, feat):
        with graph.local_scope():
            feat = self.feat_drop(feat)
            feat = self.fc(feat).view(
                -1, self._num_heads, self._out_feats)
            el = (feat * self.attn_l).sum(dim=-1).unsqueeze(-1)
            er = (feat * self.attn_r).sum(dim=-1).unsqueeze(-1)
            graph.ndata.update({'ft': feat, 'el': el})
            graph.ndata.update({'er': er})
            graph.apply_edges(fn.u_add_v('el', 'er', 'e'))
            e = self.leaky_relu(graph.edata.pop('e'))
            graph.edata['a'] = self.attn_drop(edge_softmax(graph, e))
            graph.update_all(fn.u_mul_e('ft', 'a', 'm'),
                             fn.sum('m', 'ft'))
            rst = graph.ndata['ft']
            rst = self.activation(rst)
            return rst
```

**Comment by yzh119** [comment]
It still uses message passing (i.e. `update_all`, `apply_edges`) but utilizes DGL's built-in functions (we think the combination of built-in functions could represent most graph computation patterns) instead of user-defined message/reduce functions.

Looks like you would like to deal with 2-hop neighbors, actually it's hard to handle even with customized message/reduce function. My suggestion is to create a new graph where each node is connected with it's 2-hop neighbors in the original graph, APIs like [`khop_graph`](https://docs.dgl.ai/generated/dgl.khop_graph.html?highlight=khop#dgl.khop_graph) might help.

**Comment by amblee0306** [comment]
> ```python
>  def forward(self, graph, feat):
>         with graph.local_scope():
>             feat = self.feat_drop(feat)
>             feat = self.fc(feat).view(
>                 -1, self._num_heads, self._out_feats)
>             el = (feat * self.attn_l).sum(dim=-1).unsqueeze(-1)
>             er = (feat * self.attn_r).sum(dim=-1).unsqueeze(-1)
>             graph.ndata.update({'ft': feat, 'el': el})
>             graph.ndata.update({'er': er})
>             graph.apply_edges(fn.u_add_v('el', 'er', 'e'))
>             e = self.leaky_relu(graph.edata.pop('e'))
>             graph.edata['a'] = self.attn_drop(edge_softmax(graph, e))
>             graph.update_all(fn.u_mul_e('ft', 'a', 'm'),
>                              fn.sum('m', 'ft'))
>             rst = graph.ndata['ft']
>             rst = self.activation(rst)
>             return rst
> ```

@yzh119 Thanks for the reply.
However, i m still confused. the two lines here take in the input `feat` which is the same. 
```
el = (feat * self.attn_l).sum(dim=-1).unsqueeze(-1)
er = (feat * self.attn_r).sum(dim=-1).unsqueeze(-1)
```
I expected the input to calculate el and er to be different since, the aim was to calculate (a^T [Wh_i || Wh_j] = a_l Wh_i + a_r Wh_j), where el should have been a_l Wh_i  and er should have been a_r Wh_j. Yet, feat is similar for both cases. Is there something that I am not seeing? or is it at this line below that it will put in features h_i and h_j ```graph.apply_edges(fn.u_add_v('el', 'er', 'e'))```?

**Comment by yzh119** [comment]
`feat` is the feature tensor with shape `(N,D)` that represents the feature on all nodes.

```python
el = (feat * self.attn_l).sum(dim=-1).unsqueeze(-1)
er = (feat * self.attn_r).sum(dim=-1).unsqueeze(-1)
```

These two lines compute the `a_l Wh_i` **FOR ALL NODES** i as the left hand side in the formula, and the `a_r Wh_j` **FOR ALL NODES** j as the right hand side in the formula. Note that both `feat`, `el`, `er` are tensors with leading dimension `N` that correspond to node features on all nodes, instead of a single node.

The interaction of `el` and `er` happens in 
```python
graph.apply_edges(fn.u_add_v('el', 'er', 'e'))
```
where the `fn.u_add_v('el', 'er', 'e')` describes the behavior (addition) on each edges, DGL would parallelize the computation on all edges.

I hope the following example could help you understand how does that work:

Suppose we have a graph 0->1->2->0
```
0 --> 1
 ^   /
  \ v
   2
```
Node 0 has feature `[1]`, node 1 has feature `[2]`, node 2 has feature `[3]`.
We would like to compute `3 * h_i + 2 * h_j` for all edges (i, j).

```python
>>> import torch
>>> import dgl
>>> import dgl.function as fn
>>> g = dgl.graph(([0, 1, 2], [1, 2, 0]))  # create graph
>>> feat = torch.tensor([[1.], [2.], [3.]])  # node features
>>> el = feat * 3  # 3 * h_i for all nodes as h_i
>>> er = feat * 2  # 2 * h_j for all nodes as h_j
>>> el
tensor([[3.],
        [6.],
        [9.]])
>>> er
tensor([[2.],
        [4.],
        [6.]])
>>> g.ndata.update({'el': el, 'er': er})  # copy el and er to DGLGraph's node dictionary
>>> g.apply_edges(fn.u_add_v('el', 'er', 'e'))
>>> g.edata['e']
tensor([[ 7.],
        [12.],
        [11.]])
```
which is exactly what we desired (7 = 3 * 1 + 2 * 2, 12 = 3 * 2 + 2 * 3, 11 = 3 * 3 + 2 * 1).

**Comment by yzh119** [comment]
I just found another issue from you: https://github.com/dmlc/dgl/issues/1854.
It seems you completely misunderstood our implementation...
Does the example I mentioned above helps? If not we'll seriously consider writing a better tutorial for GAT.

**Comment by amblee0306** [comment]
Hi there. sorry for the late reply. yes it is clear now. i have tried to run GAT (in the examples) that uses the gatconv module but i only got an accuracy of 81.5% as compared to the claimed 83% (is there a reason for this).

**Comment by yzh119** [comment]
Briefly speaking, we changed the cora data split in #1583 to align with the GCN's original paper. However, both pygcn and pygat (GCN/GAT's official implementation) uses the wrong data split.

This being said, you can't get an average accuracy of 83% with the correct data split. (with pygcn's data split DGL's implementation could get an average accuracy of 84%.)

btw we did not fixed the random seed in the training script so you will not get a fixed accuracy per run. Under current data split setting you can get an average test accuracy around 82% if you run the training script for multiple times.

**Comment by yzh119** [comment]
Closed because of inactivibty
